File: Cbuffer.cs
using System;


public class Cbuffer <T>
{
    private int bufferSize;
    private int tail;
    private T[] values;

	public Class1(int bufferSize)
	{
        this.bufferSize = bufferSize;
	}

    public void PushBack(double val)
    {
        // Add value to correct position
        values[tail % bufferSize] = val;

        // Update tail value
        tail++;
    }

    public AtPosition(int index)
    {
        // If trying to read value not assigned
        if (index > bufferSize)
            return;

        return values[index % bufferSize];
    }


}

File: OpenGLTemplate/CBuffer.cpp
#include "CBuffer.h"

CBuffer::CBuffer() { _bufferSize = size; }

CBuffer::CBuffer(int bufferSize) : _bufferSize(bufferSize) { tail = -1; }

CBuffer::~CBuffer() {}

// Allows value to be written into the cbuffer
void CBuffer::push_back(double val)
{
	// Add value to correct position
	values[tail % size] = val;

	// Update tail value
	tail++;
}

// Returns the value at the given index of cbuffer
double CBuffer::atPosition(int index)
{
	// If trying to read value not assigned
	if (index > tail)
		return values[tail % size];

	return values[index % size];
}

// returns the current tail of the cbuffer
int CBuffer::getTail() { return tail; }